include/bbp/sonata/output.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#include <algorithm>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <highfive/H5File.hpp>
+
+#include <bbp/sonata/population.h>
+
+namespace H5 = HighFive;
+
+namespace bbp {
+namespace sonata {
+
+// KeyType will be NodeID for somas report and pair<NodeID, uint32_t> for elemets report
+template <typename KeyType>
+struct SONATA_API DataFrame {
+    using DataType = std::map<KeyType, std::vector<float>>;
+    std::vector<double> index;
+    DataType data;
+};
+
+/**
+  const SpikeReader file(filename);
+  auto pops = file.getPopulationNames();
+  for (const auto& data: file[pops[0]].get(Selection{12UL, 34UL, 58UL})) {
+      NodeID node_id;
+      double timestamp;
+      std::tie(node_id, timestamp) = data;
+      std::cout << "[" << timestamp << "] " << node_id << std::endl;
+  }
+*/
+class SONATA_API SpikeReader
+{
+  public:
+    class Population
+    {
+      public:
+        enum class Sorting : char {
+            none = 0,
+            by_id = 1,
+            by_time = 2,
+        };
+
+        using Spike = std::pair<NodeID, double>;
+        using Spikes = std::vector<Spike>;
+        Spikes get(const Selection& node_ids = Selection({}),
+                   double tstart = -1,
+                   double tstop = -1) const;
+        Sorting getSorting() const;
+
+      private:
+        Population(const std::string& filename, const std::string& populationName);
+
+        Spikes spikes_;
+        Sorting sorting_ = Sorting::none;
+        // Use for clamping of user values
+        double tstart_, tstop_;
+
+        // Helpers to filter by node_ids
+        // Filter in place
+        void filterNode(Spikes& spikes, const Selection& node_ids) const;
+        void filterNodeIDUnsorted(Spikes& spikes, const Selection& node_ids) const;
+        void filterNodeIDSorted(Spikes& spikes, const Selection& node_ids) const;
+
+        // Helpers to filter by timestamps
+        // Filter in place
+        void filterTimestamp(Spikes& spikes, double tstart, double tstop) const;
+        void filterTimestampUnsorted(Spikes& spikes, double tstart, double tstop) const;
+        void filterTimestampSorted(Spikes& spikes, double tstart, double tstop) const;
+
+        friend SpikeReader;
+    };
+
+    explicit SpikeReader(const std::string& filename);
+
+    std::vector<std::string> getPopulationsNames() const;
+
+    const Population& getPopulation(const std::string& populationName) const;
+
+    const Population& operator[](const std::string& populationName) const;
```

Comment:
I think this was done because in the python bindings users are used to the bracket notation to refer to a population.
I don't think it hurts to have the option also in the c++ api right?

Reply:
I'd prefer to have consistency within the C++ interface for now.  If all populations are going to work like this, then we can have a separate PR updating all of them.  A grey zone where part of the API works one way, and then another works another way is annoying from a user's standpoint.